Server crashes after trying to insert a literal int into a timestamp column

Please follow the steps below to help us help you.

1. Please supply the header (i.e. the first few lines) of your most recent
   log file **for each node in your cluster**. On most unix-based systems
   running with defaults, this boils down to the output of

     grep -F '[config]' cockroach-data/logs/cockroach.INFO

```
$ grep -F '[config]' cockroach-data/logs/cockroach.INFO
I160609 10:05:46.414805 util/log/clog.go:998  [config] file created at: 2016/06/09 10:05:46
I160609 10:05:46.414805 util/log/clog.go:998  [config] running on machine: JESSEs-MacBook-Pro
I160609 10:05:46.414805 util/log/clog.go:998  [config] binary: CockroachDB beta-20160602-130-gadd8b54 (darwin amd64, built 2016/06/08 02:45:03, go1.6.2)
I160609 10:05:46.414805 util/log/clog.go:998  [config] arguments: [cockroach start]
```

2. Please describe the issue you observed:

I was trying to insert a row into this table:

```
> show create table increment;
+-----------+-------------------------------------------+
|   Table   |                CreateTable                |
+-----------+-------------------------------------------+
| increment | CREATE TABLE increment (␤                 |
|           |     a INT NOT NULL,␤                      |
|           |     b TIMESTAMP NULL DEFAULT now(),␤      |
|           |     CONSTRAINT "primary" PRIMARY KEY (a)␤ |
|           | )                                         |
+-----------+-------------------------------------------+
```

I inserted an improperly formatted value, a numeric literal, into the `TIMESTAMP` column.

```
insert into increment (b) values (2016-06-08);
driver: bad connection
```

- What did you expect to see?

I expected for it to either just work or to return an error. 

- What did you see instead?

In the SQL shell, I got `driver: bad connection`, and the node outputed the following:

```
$ cockroach start
build:     beta-20160602-130-gadd8b54 @ 2016/06/08 02:45:03 (go1.6.2)
admin:     http://localhost:8080
sql:       postgresql://root@localhost:26257?sslmode=disable
logs:      cockroach-data/logs
store[0]:  path=cockroach-data
panic: could not make constant int from literal "08" [recovered]
	panic: could not make constant int from literal "08"

goroutine 80 [running]:
panic(0x5021120, 0xc820200ed0)
	/usr/local/Cellar/go/1.6.2/libexec/src/runtime/panic.go:481 +0x3e6
github.com/cockroachdb/cockroach/sql/parser.(*Parser).Parse.func1(0xc8207932a0)
	/Users/jesseseldess/src/github.com/cockroachdb/cockroach/sql/parser/parse.go:72 +0x13a
panic(0x5021120, 0xc820200ed0)
	/usr/local/Cellar/go/1.6.2/libexec/src/runtime/panic.go:443 +0x4e9
github.com/cockroachdb/cockroach/sql/parser.(*scanner).scanNumber(0xc8202cc988, 0xc8202cca60, 0x30)
	/Users/jesseseldess/src/github.com/cockroachdb/cockroach/sql/parser/scan.go:564 +0x358
github.com/cockroachdb/cockroach/sql/parser.(*scanner).scan(0xc8202cc988, 0xc8202cca60)
	/Users/jesseseldess/src/github.com/cockroachdb/cockroach/sql/parser/scan.go:339 +0x4f8
github.com/cockroachdb/cockroach/sql/parser.(*scanner).Lex(0xc8202cc988, 0xc8202cca60, 0x0)
	/Users/jesseseldess/src/github.com/cockroachdb/cockroach/sql/parser/scan.go:81 +0x249
github.com/cockroachdb/cockroach/sql/parser.sqllex1(0x820bda8, 0xc8202cc988, 0xc8202cca60, 0xc82078f8b8, 0x0)
	yaccpar:119 +0x52
github.com/cockroachdb/cockroach/sql/parser.(*sqlParserImpl).Parse(0xc8202cca60, 0x820bda8, 0xc8202cc988, 0x0)
	yaccpar:204 +0x63f80
github.com/cockroachdb/cockroach/sql/parser.(*Parser).Parse(0xc8202cc988, 0xc8208340df, 0x2e, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/jesseseldess/src/github.com/cockroachdb/cockroach/sql/parser/parse.go:78 +0x149
github.com/cockroachdb/cockroach/sql.(*Executor).execRequest(0xc820130460, 0x72c16b8, 0xc820507680, 0xc8202cc800, 0xc8208340df, 0x2e, 0x0, 0x0, 0x0, 0x8211600)
	/Users/jesseseldess/src/github.com/cockroachdb/cockroach/sql/executor.go:377 +0xdc
github.com/cockroachdb/cockroach/sql.(*Executor).ExecuteStatements(0xc820130460, 0x72c16b8, 0xc820507680, 0xc8202cc800, 0xc8208340df, 0x2e, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/jesseseldess/src/github.com/cockroachdb/cockroach/sql/executor.go:338 +0xe5
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).executeStatements(0xc82061ef00, 0x72c16b8, 0xc820507680, 0xc8208340df, 0x2e, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/jesseseldess/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:648 +0xb2
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).handleSimpleQuery(0xc82061ef00, 0x72c16b8, 0xc820507680, 0xc82061ef28, 0x0, 0x0)
	/Users/jesseseldess/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:325 +0xf2
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).serve(0xc82061ef00, 0x0, 0x0, 0x0)
	/Users/jesseseldess/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:281 +0x1095
github.com/cockroachdb/cockroach/sql/pgwire.(*Server).ServeConn(0xc82014e600, 0x82146a8, 0xc8200e8580, 0x0, 0x0)
	/Users/jesseseldess/src/github.com/cockroachdb/cockroach/sql/pgwire/server.go:226 +0x9e5
github.com/cockroachdb/cockroach/server.(*Server).Start.func8.1(0x82146a8, 0xc8200e8580)
	/Users/jesseseldess/src/github.com/cockroachdb/cockroach/server/server.go:362 +0x42
github.com/cockroachdb/cockroach/util.(*Server).ServeWith.func1(0xc820124120, 0x82146a8, 0xc8200e8580, 0xc820346000)
	/Users/jesseseldess/src/github.com/cockroachdb/cockroach/util/net.go:131 +0x62
created by github.com/cockroachdb/cockroach/util.(*Server).ServeWith
	/Users/jesseseldess/src/github.com/cockroachdb/cockroach/util/net.go:133 +0x333
```

--- a/sql/parser/scan.go
+++ b/sql/parser/scan.go
@@ -548,7 +548,9 @@ func (s *scanner) scanNumber(lval *sqlSymType, ch int) {
 		lval.id = FCONST
 		floatConst := constant.MakeFromLiteral(lval.str, token.FLOAT, 0)
 		if floatConst.Kind() == constant.Unknown {
-			panic(fmt.Sprintf("could not make constant float from literal %q", lval.str))
+			lval.id = ERROR
+			lval.str = fmt.Sprintf("could not make constant float from literal %q", lval.str)
+			return
 		}
 		lval.union.val = &NumVal{Value: floatConst, OrigString: lval.str}
 	} else {
@@ -561,7 +563,9 @@ func (s *scanner) scanNumber(lval *sqlSymType, ch int) {
 		lval.id = ICONST
 		intConst := constant.MakeFromLiteral(lval.str, token.INT, 0)
 		if intConst.Kind() == constant.Unknown {
-			panic(fmt.Sprintf("could not make constant int from literal %q", lval.str))
+			lval.id = ERROR
+			lval.str = fmt.Sprintf("could not make constant int from literal %q", lval.str)
+			return
 		}
 		lval.union.val = &NumVal{Value: intConst, OrigString: lval.str}
 	}

--- a/sql/parser/scan_test.go
+++ b/sql/parser/scan_test.go
@@ -396,6 +396,7 @@ func TestScanError(t *testing.T) {
 		{`1.0x`, "invalid hexadecimal literal"},
 		{`0x0x`, "invalid hexadecimal literal"},
 		{`00x0x`, "invalid hexadecimal literal"},
+		{`08`, "could not make constant int from literal \"08\""},
 		{`$9223372036854775809`, "integer value out of range"},
 	}
 	for _, d := range testData {

